runtime/vm/OutOfLineINL_jdk_internal_foreign_abi_ProgrammableInvoker.cpp

```diff
@@ -0,0 +1,162 @@
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "VMHelpers.hpp"
+#include "BytecodeAction.hpp"
+#include "UnsafeAPI.hpp"
+#include "j9vmnls.h"
+#include "omrlinkedlist.h"
+#include "OutOfLineINL.hpp"
+#include "FFITypeHelpers.hpp"
+#include "AtomicSupport.hpp"
+
+extern "C" {
+
+#if JAVA_SPEC_VERSION >= 16
+/* jdk.internal.foreign.abi.ProgrammableInvoker: private static synchronized native void resolveRequiredFields(); */
+VM_BytecodeAction
+OutOfLineINL_jdk_internal_foreign_abi_ProgrammableInvoker_resolveRequiredFields(J9VMThread *currentThread, J9Method *method)
+{
+	VM_BytecodeAction rc = EXECUTE_BYTECODE;
+	J9JavaVM *vm = currentThread->javaVM;
+	J9ConstantPool *jclConstantPool = (J9ConstantPool *)vm->jclConstantPool;
+	const int cpEntryNum = 2;
+	U_16 cpIndex[cpEntryNum] = {
+				J9VMCONSTANTPOOL_JDKINTERNALFOREIGNABIPROGRAMMABLEINVOKER_CIFNATIVETHUNKADDR,
+				J9VMCONSTANTPOOL_JDKINTERNALFOREIGNABIPROGRAMMABLEINVOKER_ARGTYPESADDR
+			};
+
+	for (int i = 0; i < cpEntryNum; i++) {
+		J9RAMFieldRef *cpFieldRef = ((J9RAMFieldRef*)jclConstantPool) + cpIndex[i];
+		UDATA const flags = cpFieldRef->flags;
+		UDATA const valueOffset = cpFieldRef->valueOffset;
+
+		if (!VM_VMHelpers::instanceFieldRefIsResolved(flags, valueOffset)) {
+			resolveInstanceFieldRef(currentThread, NULL, jclConstantPool, cpIndex[i], J9_RESOLVE_FLAG_NO_THROW_ON_FAIL | J9_RESOLVE_FLAG_JCL_CONSTANT_POOL, NULL);
+			if (VM_VMHelpers::exceptionPending(currentThread)) {
+				rc = GOTO_THROW_CURRENT_EXCEPTION;
+				goto done;
+			}
+		}
+	}
+
+done:
+	VM_OutOfLineINL_Helpers::returnVoid(currentThread, 0);
+	return rc;
+}
+
+/* jdk.internal.foreign.abi.ProgrammableInvoker: private native void initCifNativeThunkData(String[] argLayouts, String retLayout, boolean newArgTypes); */
+VM_BytecodeAction
+OutOfLineINL_jdk_internal_foreign_abi_ProgrammableInvoker_initCifNativeThunkData(J9VMThread *currentThread, J9Method *method)
```

Comment:
can you add a comment header briefly describing whats happening here. Also, describe the java args.

Reply:
Will update accordingly soon.

Reply:
Updated the comment as suggested above.